### PR TITLE
Non-multilevel Optimizers reworked

### PR DESCRIPTION
**Pull request by PaulTalbot-INL**
Fixes all the non-multilevel optimizers for the new dataobject rework.

I'll finish the multilevel in another PR.

Adds a "writeSteps" option to the "initialization" block of the Optimizer.  This defaults to "every", in which case every optimizer iteration is written to the SolutionExport.  If instead it is set to "final", ONLY the most optimal point is written to the SolutionExport.

### CODE REVIEW

**Comment by PaulTalbot-INL** [comment]
Running the test differences now....

**Comment by PaulTalbot-INL** [comment]
Test results.  I did add one test to cover the new features specifically.

```
**** THE FOLLOWING TESTS CHANGED OUTCOMES ****
**** Test framework/Optimizers.Beale | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/Optimizers.BealeFailureRejection | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/Optimizers.BealeFiniteDifference | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/Optimizers.BealeWithSampler | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/Optimizers.Boundary | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/Optimizers.Constraint | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/Optimizers.GoldsteinPrice | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/Optimizers.Max | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/Optimizers.McCormick | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/Optimizers.Persistence | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/Optimizers.Stochastic | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/Optimizers.StochasticFiniteDifference | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/Optimizers.TrajGrid | has changed outcome from <FAILED> to <OK> !!!!
**** Test framework/Optimizers.Verbose | has changed outcome from <FAILED> to <OK> !!!!
*****************
**** THE FOLLOWING TESTS HAVE BEEN ADDED IN 2ND FILE ****
**** Test framework/Optimizers.BealeFinal | has been ADDED  ! It is present in 2nd file BUT NOT in 1st file !!!!
*****************
Number of success first file:
383
Number of success second file:
397
NUMBER OF SUCCESSES INCREASED: +14
**** NUMBER OF TESTS IN THE TWO FILES ARE DIFFERENT: ****
**** first file :532 /= second file: 533****
TOTAL TIME OF TESTS INCREASED       : reference_time(511.4) < new_time(514.3)
AVERAGE TIMES FOR SUCCESS TESTS ARE : reference_time(1.33524804178) v.s. new_time(1.29546599496)
*********************************
```

File: framework/Optimizers/SPSA.py [hunk 1]
```diff
@@ -119,7 +119,7 @@ def _newOptPointAdd(self, gradient, traj):
     """
     ak = self._computeGainSequenceAk(self.paramDict,self.counter['varsUpdate'][traj],traj) # Compute the new ak
     self.optVarsHist[traj][self.counter['varsUpdate'][traj]] = {}
-    varK = copy.deepcopy(self.counter['recentOptHist'][traj][0]['inputs'])
+    varK = dict((var,self.counter['recentOptHist'][traj][0][var]) for var in self.getOptVars(traj)) #copy.deepcopy(self.counter['recentOptHist'][traj][0]['inputs'])
```
Comment (wangcj05) [inline framework/Optimizers/SPSA.py]: rm commented part?

File: tests/framework/Optimizers/beale.xml [hunk 2]
```diff
@@ -9,12 +9,18 @@
       This test runs the optimization on Beale's function.  It tests analytic optimization values
       as well as the mechanical operation of the test.  Also tests the "gainGrowthFactor" and
       "gainShrinkFactor" convergence parameters.
+
```
Comment (wangcj05) [inline tests/framework/Optimizers/beale.xml]: remove empty line?

Comment (PaulTalbot-INL) [reply tests/framework/Optimizers/beale.xml]: I think it improves readability.

File: framework/Optimizers/SPSA.py (hunk 1 above)
Comment (PaulTalbot-INL) [reply framework/Optimizers/SPSA.py]: I want to leave this part there just in case the deepcopy ends up coming up again, so I can find it easily.

Comment (wangcj05) [reply framework/Optimizers/SPSA.py]: ok

File: tests/framework/Optimizers/beale.xml (hunk 2 above)
Comment (wangcj05) [reply tests/framework/Optimizers/beale.xml]: ok

**Comment by wangcj05** [comment]
@PaulTalbot-INL Please share your new reference test results with raven developers. Thanks.

**Pull request merged**